lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,474 @@
+            {
+                Exceptions.Throw(new IMRUSystemException("masterTaskId cannot be null"), Logger);
+            }
+
+            _totalExpectedTasks = numTasks;
+            _masterTaskId = masterTaskId;
+        }
+
+        /// <summary>
+        /// Adds a Task to the task collection
+        /// Throws IMRUSystemException in the following cases:
+        ///   taskId is already added 
+        ///   taskConfiguration is null
+        ///   activeContext is null
+        ///   trying to add extra tasks
+        ///   trying to add Master Task twice
+        ///   No Master Task is added in the collection
+        /// </summary>
+        /// <param name="taskId"></param>
+        /// <param name="taskConfiguration"></param>
+        /// <param name="activeContext"></param>
+        internal void AddTask(string taskId, IConfiguration taskConfiguration, IActiveContext activeContext)
+        {
+            if (taskId == null)
+            {
+                Exceptions.Throw(new IMRUSystemException("The taskId is null."), Logger);
+            }
+
+            if (_tasks.ContainsKey(taskId))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] already exists.", taskId);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (taskConfiguration == null)
+            {
+                Exceptions.Throw(new IMRUSystemException("The task configuration is null."), Logger);
+            }
+
+            if (activeContext == null)
+            {
+                Exceptions.Throw(new IMRUSystemException("The context is null."), Logger);
+            }
+
+            if (NumberOfTasks >= _totalExpectedTasks)
+            {
+                string msg = string.Format("Trying to add an additional Task {0}, but the total expected Task number {1} has been reached.", taskId, _totalExpectedTasks);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (taskId.Equals(_masterTaskId) && MasterTaskExists())
+            {
+                string msg = string.Format("Trying to add second master Task {0}.", taskId);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            _tasks.Add(taskId, new TaskInfo(new TaskStateMachine(), taskConfiguration, activeContext));
+
+            if (NumberOfTasks == _totalExpectedTasks && !MasterTaskExists())
+            {
+                Exceptions.Throw(new IMRUSystemException("There is no master task added."), Logger);
+            }
+        }
+
+        /// <summary>
+        /// Returns the number of tasks in the task collection
+        /// </summary>
+        internal int NumberOfTasks
+        {
+            get { return _tasks.Count; }
+        }
+
+        /// <summary>
+        /// This method is called when receiving IRunningTask event during system SubmittingTasks state.
```

Comment:
We don't check system state in this code. Will driver code do this check? I'm not sure we should mention system states in the comments for these methods if this code doesn't actually check for validity of getting this event in this system state.